packages/utilities/src/inferLanguages.ts

```diff
@@ -0,0 +1,122 @@
+import {
+    Concept,
+    Containment,
+    Language,
+    Link,
+    Property,
+    Reference,
+    SerializationChunk,
+    builtinPrimitives,
+    chain,
+    concatenator,
+    lastOf
+} from "@lionweb/core"
+import { hasher } from "./hashing.js"
+
+const id = chain(concatenator("-"), hasher())
+const key = lastOf
+
+const { stringDatatype } = builtinPrimitives
+
+export function inferLanguagesFromChunk(chunk: SerializationChunk): Language[] {
+    const languages = new Map<string, Language>()
+    const concepts = new Map<string, Concept>()
+    const links = new Array<{ link: Link; conceptId: string }>()
+
+    for (const chunkLanguage of chunk.languages) {
+        const languageName = chunkLanguage.key
+        const language = new Language(languageName, chunkLanguage.version, id(languageName), key(languageName))
+        languages.set(languageName, language)
+    }
+
+    for (const node of chunk.nodes) {
+        const languageName = node.classifier.language
+        const entityName = node.classifier.key
+
+        const language = getLanguage(languages, languageName)
+        if (language.entities.filter(entity => entity.key === entityName).length) {
+            continue
+        }
+
+        const concept = new Concept(language, entityName, key(language.name, entityName), id(language.name, entityName), false)
+        language.havingEntities(concept)
+        concepts.set(node.id, concept)
+
+        for (const property of node.properties) {
+            const propertyName = property.property.key.split("_")[2]
+            if (concept.features.filter(feature => feature.key === propertyName).length) {
+                continue
+            }
+
+            // TODO: Infer the property type from the value
+            const feature = new Property(
+                concept,
+                propertyName,
+                key(languageName, concept.name, propertyName),
+                id(languageName, concept.name, propertyName)
+            )
+                .ofType(stringDatatype)
+                .havingKey(property.property.key)
+
+            concept.havingFeatures(feature)
+        }
+
+        for (const containment of node.containments) {
+            const containmentName = containment.containment.key
+
+            if (concept.features.filter(feature => feature.key === containmentName).length) {
+                continue
+            }
+
+            // TODO: Check if the containment is multiple or single
+            const feature = new Containment(
+                concept,
+                containmentName,
+                key(languageName, concept.name, containmentName),
+                id(languageName, concept.name, containmentName)
+            ).isMultiple()
+            concept.havingFeatures(feature)
+
+            // TODO: What if `children` list is empty?
```

Comment:
Should never be empty according to spec. But you could also use the `asArray` function from `@lionweb/core`.